staging/src/k8s.io/api/bulk/v1alpha1/types.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient:nonNamespaced
+
+// BulkRequest contains a single request to bulk api.
+type BulkRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Opaque identifier used to match requests and responses.
+	RequestID string `json:"requestId" protobuf:"bytes,1,name=requestId"`
+
+	// Starts new watch.
+	// +optional
+	Watch *WatchOperation `json:"watch,omitempty" protobuf:"bytes,2,opt,name=watch"`
+
+	// Starts new watch.
+	// +optional
+	WatchList *WatchListOperation `json:"watchList,omitempty" protobuf:"bytes,3,opt,name=watchList"`
+
+	// Stops existing watch.
+	// +optional
+	StopWatch *StopWatchOperation `json:"stopWatch,omitempty" protobuf:"bytes,4,opt,name=stopWatch"`
+
+	// Gets list of items.
+	// +optional
+	List *ListOperation `json:"list,omitempty" protobuf:"bytes,5,opt,name=list"`
+
+	// Gets single item.
+	// +optional
+	Get *GetOperation `json:"get,omitempty" protobuf:"bytes,6,opt,name=get"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient:nonNamespaced
+
+// BulkResponse contains a single response from bulk api.
+type BulkResponse struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Opaque identifier propagated from the request.
+	// Null for messages issued by server (WatchEvent, WatchStopped triggered by server)
+	// +optional
+	RequestID *string `json:"requestId,omitempty" protobuf:"bytes,1,opt,name=requestId"`
+
+	// Contains error details when request was failed.
+	// +optional
+	Failure *metav1.Status `json:"failure,omitempty" protobuf:"bytes,2,opt,name=failure"`
```

Comment:
We named this `Result` in admission web hooks.  We should be consistent with that pattern unless we have a reason.

Reply:
This field acts as 'type-of-response' switch (fields Failure/WatchStopped/WatchStarted/... are mutually exclusive).
So it is some sort of "event-type" (corresponds to "previous request was failed").
IMO event/field name should clearly indicate unexpected error.

Also similar field is named 'ResponseStatus' in audit api 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/audit/types.go#L110

It looks much better for our case. Alternatively we can create submessage like:
   type OperationFailed struct {    // ResponseFailed?
      Result *metav1.Status
   }

And use OperationFailed instead of metav1.Status
   type BulkResponse {
     // ...
     OperationFailed *OperationFailed
     // ...
   }

Reply:
Why the union rather than just having the possibility of sending multiple different messages? E.g. the ID field doesn't make sense for everything, at least per the comment on it.

Reply:
Without union it is harder to serialize/deserialize messages. Some dynamic parsing (like json.RawMessage) might be required do determine type of object.  Using of protobufs is also  complicated. So
 - is it needed to create new serialization/deserialization logic, which might cause some divergences with traditional RESP apis (hidden bugs);
 - any tricks around RawMessage introduce additional overhead.